File: 72.cpp
struct Solution {
    int minDistance(const string &word1, const string &word2) {
        if (word1.size() == 0 || word2.size() == 0){
            return word1.size() + word2.size();
        }
        vector<int> dp(word2.size(), 0);
        dp[0] = (word1[0] == word2[0] ? 0 : 1);
        for (int j = 1; j < word2.size(); ++j){
            dp[j] = (word1[0] == word2[j] ? j : 1 + min(j, dp[j - 1]));
        }
        int topleft, temp;
        for (int i = 1; i < word1.size(); ++i){
            topleft = dp[0];
            dp[0] = (word1[i] == word2[0] ? i : 1 + min(i, topleft));
            for (int j = 1; j < word2.size(); ++j){
                temp = topleft;
                topleft = dp[j];
                dp[j] = (word1[i] == word2[j] ? temp : 1 + min(temp, min(dp[j - 1], dp[j])));
            }
        }
        return dp.back();
    }
};

File: sol/72.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        if len(word1) == 0 or len(word2) == 0:
            return len(word1) + len(word2)
        dp = [0] * len(word2)
        dp[0] = 0 if word1[0] == word2[0] else 1
        for j in range(1, len(word2)):
            dp[j] = j if word1[0] == word2[j] else (1 + min(dp[j - 1], j))
        for i in range(1, len(word1)):
            topleft = dp[0]
            dp[0] = i if word1[i] == word2[0] else (1 + min(dp[0], i))
            for j in range(1, len(word2)):
                topleft, dp[j] = dp[j], topleft if word1[i] == word2[j] else (1 + min(topleft, dp[j - 1], dp[j]))
        return dp[-1]
